File: 剑指offer/45.把数组排成最小的数.py
from typing import List


class Solution:
    def minNumber(self, nums: List[int]) -> str:
        def quick_sort(nums, left, right):
            if left >= right:
                return

            i = left
            j = right

            privot = nums[left]

            while i < j:
                while i < j and nums[j] + privot > privot + nums[j]:
                    j -= 1
                nums[i] = nums[j]

                while i < j and nums[i] + privot <= privot + nums[i]:
                    i += 1
                nums[j] = nums[i]

            nums[i] = privot

            quick_sort(nums, left, i - 1)
            quick_sort(nums, i + 1, right)

        nums = [str(num) for num in nums]
        quick_sort(nums, 0, len(nums) - 1)
        return ''.join(nums)

File: 剑指offer/45.把数组排成最小的数.cpp
#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    static bool compare(const string& s1, const string& s2)
    {
        string ab = s1 + s2;
        string ba = s2 + s1;
        return ab < ba; //升序排列。如改为ab > ba, 则为降序排列
    }

    string minNumber(vector<int>& nums)
    {
        vector<string> strs;
        for (int num : nums) {
            strs.push_back(to_string(num));
        }

        sort(strs.begin(), strs.end(), compare);

        string ans = "";
        for (string num : strs) {
            ans.append(num);
        }

        return ans;
    }
};